crates/iota-data-ingestion-core/src/tests.rs

```diff
@@ -253,6 +277,45 @@ async fn worker_pool_with_reducer() {
     assert_eq!(result.unwrap().get("test"), Some(&20));
 }
 
+// Tests the graceful shutdown behavior when reducer encounter persistent
+// failures.
+//
+// This test verifies that:
+// 1. When Reducer::commit implementation continuously fails.
+// 2. The exponential backoff retry mechanism would normally create an loop
+//    until the successful value is returned.
+// 3. The graceful shutdown logic successfully breaks these retry loops upon
+//    cancellation.
+// 4. The Reducer exit cleanly without committing any batch.
+//
+// The test uses `FaultyReducer` which always fails, simulating a worst-case
+// scenario where all WorkerPools are unable to send progress data to
+// IndexerExecutor.
+#[tokio::test]
+async fn graceful_shutdown_faulty_reducer() {
+    // create a reducer with max batch of 5
+    let reducer = FaultyReducer::new(5);
+    let mut bundle = create_executor_bundle().await;
+    // Create worker pool with reducer
+    let pool = WorkerPool::new_with_reducer(TestWorker, "test".to_string(), 5, reducer);
+    bundle.executor.register(pool).await.unwrap();
+
+    let path = temp_dir();
+    for checkpoint_number in 0..20 {
+        let bytes = mock_checkpoint_data_bytes(checkpoint_number);
+        std::fs::write(path.join(format!("{}.chk", checkpoint_number)), bytes).unwrap();
+    }
+    let result = run(
+        bundle.executor,
+        Some(path),
+        Some(Duration::from_secs(1)),
+        bundle.token,
```

Comment:
Who is triggering the cancel token in this test?

Reply:
in the `run` function: https://github.com/iotaledger/iota/blob/develop/crates/iota-data-ingestion-core/src/tests.rs#L72
if you provide the `Duration` as `Some` it will wait the indicated duration and cancel the token afterwards. So in this case it will cancel the token after 1 second